Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -364,6 +364,12 @@ static ErrorOr<bool> read_eob(Scan& scan, u32 symbol)
 
         scan.end_of_bands_run_count = additional_value + (1 << eob_base) - 1;
 
+        if (scan.end_of_bands_run_count >= 1) {
+            // end_of_bands_run_count is decremented at the end of `build_macroblocks`.
+            // This line compensate this effect.
+            ++scan.end_of_bands_run_count;
+        }
```

Comment:
This logic is updated again in a later commit. Within a single PR, it's fine to update the logic at once instead of splitting it in two commits - I don't think there's a tangible benefit in doing so.

Reply:
> This feels like it could very easily break apart in the future...

I agree that it's not the best, but there are several things I want to be sure you have in mind:

1. Missing that part will result in huge failures, not a difference in a simple pixel.
2. `end_of_bands_run_count` is only written in this function, so it's pretty easy to bisect if someone has the issue. And I tried to make it clear with a commit dedicated to this.

> Could you move the logic to build_macroblocks instead, by e.g. inspecting the return value? (This function already returns true in this case)

`read_eob` is called in `read_ac` so it means that `read_ac` will need to return `end_of_bands_run_count` which doesn't make much sense IMO. And it also means another variable in the `add_ac` function to handle if we have seen an eob in the current block (if we reached an EOB0).
I'm not sure that we want that too, it's already difficult to understand that for loop, no need to complexify it.

> This logic is updated again in a later commit. Within a single PR, it's fine to update the logic at once instead of splitting it in two commits - I don't think there's a tangible benefit in doing so.

It's probably an issue on my side, I will ensure that it works and if it's okay I will put the second version in the first commit.

Reply:
> It's probably an issue on my side, I will ensure that it works and if it's okay I will put the second version in the first commit.

Actually, this is not possible, so I squashed both commits